File: output/xCodeEval_remove_8000/C++/C/7b2ed196ac86d9c6a19962e13d67ff47.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define MAX_N 100
#define MAX_K 12

typedef long long ll;

ll max(ll a, ll b) {
    return (a > b) ? a : b;
}

ll sum(int arr[], int size) {
    ll s = 0;
    for (int i = 0; i < size; i++) {
        s += arr[i];
    }
    return s;
}

ll rec(int l, int r, int k, int arrs[][MAX_K], ll sums[], ll dps[][MAX_K]) {
    ll ans = 0;
    if (l + 1 == r) {
        ans = dps[k][MAX_K - 1];
        ll sum = 0;
        for (int i = 0; i < k && i < MAX_K; i++) {
            sum += arrs[l][i];
            ans = max(ans, dps[k - i - 1][MAX_K - 1] + sum);
        }
    } else {
        int m = (l + r) / 2;
        for (int i = 0; i <= k; i++) {
            for (int j = 0; j <= k; j++) {
                dps[i][j] = dps[i][j + 1];
            }
        }
        for (int i = l; i < m; i++) {
            ll v = sums[i];
            int k1 = sizeof(arrs[i]) / sizeof(arrs[i][0]);
            for (int j = k; j >= k1; j--) {
                dps[k][j] = max(dps[k][j], dps[k][j - k1] + v);
            }
        }
        ans = max(ans, rec(m, r, k, arrs, sums, dps));
        for (int i = 0; i <= k; i++) {
            for (int j = 0; j <= k; j++) {
                dps[i][j] = dps[i][j + 1];
            }
        }
        for (int i = m; i < r; i++) {
            ll v = sums[i];
            int k1 = sizeof(arrs[i]) / sizeof(arrs[i][0]);
            for (int j = k; j >= k1; j--) {
                dps[k][j] = max(dps[k][j], dps[k][j - k1] + v);
            }
        }
        ans = max(ans, rec(l, m, k, arrs, sums, dps));
    }
    return ans;
}

void solve() {
    int n, k;
    scanf("%d %d", &n, &k);
    int arrs[MAX_N][MAX_K];
    ll sums[MAX_N];
    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        for (int j = 0; j < t; j++) {
            scanf("%d", &arrs[i][j]);
        }
        if (t > k) {
            t = k;
        }
        sums[i] = sum(arrs[i], t);
    }
    ll dps[MAX_K][MAX_K];
    memset(dps, 0, sizeof(dps));
    printf("%lld\n", rec(0, n, k, arrs, sums, dps));
}

int main() {
    solve();
    return 0;
}

File: output/xCodeEval_remove_8000/C++/Go/7b2ed196ac86d9c6a19962e13d67ff47.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func sum(arr []int64) int64 {
	var s int64
	for _, num := range arr {
		s += num
	}
	return s
}

func rec(l, r, k int, dps [][]int64, sums []int64, arrs [][]int64) int64 {
	if l+1 == r {
		ans := dps[k][k]
		sum := int64(0)
		for i := 0; i < min(k, len(arrs[l])); i++ {
			sum += arrs[l][i]
			ans = max(ans, dps[k][k-i-1]+sum)
		}
		return ans
	} else {
		m := (l + r) / 2
		dps1 := make([][]int64, len(dps))
		for i := range dps {
			dps1[i] = make([]int64, len(dps[i]))
			copy(dps1[i], dps[i])
		}
		for i := l; i < m; i++ {
			v := sums[i]
			k1 := len(arrs[i])
			for j := k; j >= k1; j-- {
				dps1[j] = append(dps1[j], dps1[j-k1]+v)
			}
		}
		ans := rec(m, r, k, dps1, sums, arrs)
		for i := range dps {
			copy(dps[i], dps1[i])
		}
		for i := m; i < r; i++ {
			v := sums[i]
			k1 := len(arrs[i])
			for j := k; j >= k1; j-- {
				dps[j] = append(dps[j], dps[j-k1]+v)
			}
		}
		ans = max(ans, rec(l, m, k, dps, sums, arrs))
		return ans
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func solve(reader io.Reader, writer io.Writer) {
	var n, k int
	fmt.Fscan(reader, &n, &k)
	sums := make([]int64, n)
	arrs := make([][]int64, n)
	for i := 0; i < n; i++ {
		var t int
		fmt.Fscan(reader, &t)
		arrs[i] = make([]int64, t)
		for j := 0; j < t; j++ {
			fmt.Fscan(reader, &arrs[i][j])
		}
		if t > k {
			arrs[i] = arrs[i][:k]
		}
		sums[i] = sum(arrs[i])
	}
	dps := make([][]int64, 13)
	for i := range dps {
		dps[i] = make([]int64, k+1)
	}
	fmt.Fprint(writer, rec(0, n, k, dps, sums, arrs))
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	solve(reader, writer)
	writer.Flush()
}
